engine/src/main/java/org/terasology/rendering/nui/layers/ingame/metrics/DebugMetricsSystem.java

```diff
@@ -0,0 +1,137 @@
+ * obtain the mode currently pointed at. Further modes can be added to the set via the register method, while registered
+ * modes can be removed via the unregister method. A convenience method unregisterAll() removes all registered modes at
+ * once. MetricsModes are currently used by the DebugOverlay instance, displaying runtime statistics on screen.
+ * unregister and unregisterAll functions.
+ */
+@RegisterSystem
+@Share(DebugMetricsSystem.class)
+public class DebugMetricsSystem extends BaseComponentSystem {
+
+    private final MetricsMode defaultMode = new NullMetricsMode();
+    private ArrayList<MetricsMode> modes;
+    private MetricsMode currentMode;
+    private int cursor;
+
+    @Override
+    public void initialise() {
+        modes = new ArrayList<>();
+
+        register(defaultMode);
+        register(new RunningMeansMode());
+        register(new SpikesMode());
+        register(new AllocationsMode());
+        register(new RunningThreadsMode());
+        register(new WorldRendererMode());
+        register(new RenderingExecTimeMeansMode("Rendering - Execution Time: Running Means - Sorted Alphabetically"));
+        currentMode = defaultMode;
+    }
+
+
+    /**
+     * Adds a MetricsMode instance to the set. Use the toggle() and getCurrentMode() methods to iterate over the set and
+     * obtain the MetricsMode instances.
+     * @param mode a MetricsMode instance
+     */
+    public boolean register(MetricsMode mode) {
+        Preconditions.checkNotNull(mode, "Illegal argument passed, mode is null.");
+
+        if (!modes.contains(mode)) {
+            return modes.add(mode);
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Returns current mode, initializes a default {@link NullMetricsMode}, if currentMode is null
+     * @return the current MetricsMode instance
+     */
+    public MetricsMode getCurrentMode() {
+        return currentMode;
+    }
+
+    /**
+     * Iterates through the set of registered MetricsMode and returns the first instance whose method isAvailable()
+     * returns true. Notice that this could be the MetricsMode that was current at the time this method was invoked.
+     * @returns a MetricsMode instance
+     */
+    public MetricsMode toggle() {
+        do {
+            cursor = (cursor + 1) % modes.size();
+            currentMode = modes.get(cursor);
+        } while (!currentMode.isAvailable());
+
+        return currentMode;
+    }
+
+
+    /**
+     * Removes from the set the MetricsMode instance provided as input.
+     *
+     * If the MetricsMode instance is the mode currently pointed at by the iterator, toggles the iterator forward. Unregistering
+     * defaultMode is not allowed, therefore throws {@link IllegalArgumentException}.
+     *
+     * @return True if the MetricsMode instance was in the set. False otherwise.
+     */
+    public boolean unregister(MetricsMode mode) {
```

Comment:
In the previous comment regarding this method I suggested to add to the javadoc the exceptions this method throws. There's also this form of method definition that includes the exceptions thrown:

`public boolean unregister(MetricsMode mode) throws except1, except2 {`

I'm not sure when to do it only in the javadoc, when to do it only in the method definition and when to do it both ways. @msteiger: Maaaaaaaaaartin?!?!?!?

By the way, the same applies to the register method.